docs/weave/weave-api-spec/04-querying.md

```diff
@@ -0,0 +1,143 @@
+
+Some available bucket paths: `/wallets`, `/auth`, `/aswaps` ...
+
+## Indexes
+
+### Primary indexes
+
+For accessing the data resides inside buckets, indexes are used. E.g. to access wallet with primary index `00CAFE00`**(hex)**, call has to be made to `/wallets` path with index as data.
+
+[//]: # 'TODO change testnet URLs to mainnet after it is launched'
+
+```bash
+curl -X POST -d '{ "json-rpc": 2.0, "id": "foobar321", "method": "abci_query", "params": { "path": "/wallets", "data": "CBC76ADED2C9DB439DB4C8D714CF26DAE5229A91" } }' https://bns.kissnet.iov.one/
+```
+
+- Path: `/`, Data: `0123456789` (hex) -> db.Get(`0123456789`)
+
+  - Queries made to root (`/`) are direct queries to the key-value DB without an index.
+
+- Path: `/wallets`, Data: `00CAFE00` (hex) -> wallets.Get(`00CAFE00`)
+  - `wallets` are queried for the account with address `00CAFE00`.
+
+### Secondary indexes
+
+Another way to access data is using **secondary indexes**. Via secondary indexes, data could have multiple ways to be accessed. E.g. wallets are registered under a name so there might be some use cases for accessing them with names. `/wallets` + `/name` = `/wallets/name` indicates that bucket will be queried using name index that will be sent in the data field.
+
+- Path: `/wallets/name`, Data: `4A6F686E` (raw: `John`): wallets.Index("name").Get("4A6F686E")
+  - `wallets` are queried for the account with name `John`.
+
+### Multikey indexes
+
+There might be some cases where one index has multiple values. Multikeys exists for this purpose. Query response will be multiple values instead of one.
+
+### Prefixes
+
+There might be cases where all the data with has the same index prefix wanted to be retrieved. For this purpose you can do prefix searches using `?prefix` endpoints
+
+- Queries that made using prefix without a data field would result as listing all the objects saved under the bucket
+
+  - E.g. query below will return all the wallets saved under wallet bucket in [ResultSet](#Responses) response format
+
+    ```bash
+    curl -X POST -d '{ "json-rpc": 2.0, "id": "foobar321", "method": "abci_query",
+    "params": { "path": "/wallets?prefix", "data": "" } }' \
+    https://bns.kissnet.iov.one/
+    ```
+
+  - E.g. query below will return all the tokens saved under token bucket in again ResultSet.
+
+    ```bash
+      curl -X POST -d '{ "json-rpc": 2.0, "id": "foobar321", "method": "abci_query",
+      "params": { "path": "/tokens?prefix", "data": "" } }' \
+      https://bns.kissnet.iov.one/
+    ```
+
+- Path: `/wallets?prefix`, Data: `0123456789` (hex) -> db.Iterator(`0123456789`, `012345678A`)
+  - `wallets` are queried for the accounts starting with from `0123456789` to `012345678A`
+
+## Responses
+
+Since Weave queries routed to tendermint ABCI protocol, tendermint enforces responses to be in `key/value` format. Reference: [tendermint/abci-spec#query](https://tendermint.readthedocs.io/en/v0.21.0/abci-spec.html#query).
```

Comment:
```suggestion
Since Weave queries are routed to tendermint ABCI protocol, tendermint forces responses to be in `key/value` format. Reference: [tendermint/abci-spec#query](https://tendermint.readthedocs.io/en/v0.21.0/abci-spec.html#query).
```